### PR TITLE
Add strict type for Port, Host and URL

### PR DESCRIPTION
**Pull request by balamurugana**
New subpackage `net` has types
* Port - extended uint16 native type contains friendly methods and functions.
* Host - A type holds host and Port pairs where Port is optional.
* URL - extended url.URL native type which gives more validation and JSON friendly.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: cmd/endpoint.go [hunk 1]
```diff
@@ -43,7 +44,7 @@ const (
 
 // Endpoint - any type of endpoint.
 type Endpoint struct {
-	*url.URL
+	*url2.URL
```
Comment (harshavardhana) [inline cmd/endpoint.go]: Why not use `url.URL`  and name the `url `instance as `gourl` ?

File: pkg/url/url.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
+package url
```
Comment (harshavardhana) [inline pkg/url/url.go]: Requires license header and some tests here..

File: pkg/url/url.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+// UnmarshalJSON - parses given data into URL.
+func (u *URL) UnmarshalJSON(data []byte) (err error) {
+	var s string
+	if err = json.Unmarshal(data, &s); err != nil {
+		return err
+	}
+
+	var ru *URL
+	if ru, err = Parse(s); err != nil {
+		return err
```
Comment (harshavardhana) [inline pkg/url/url.go]: :cake:

File: cmd/endpoint.go (hunk 1 above)
Comment (balamurugana) [reply cmd/endpoint.go]: can be done :+1:

File: cmd/endpoint.go [hunk 4]
```diff
@@ -18,16 +18,15 @@ package cmd
 
 import (
 	"fmt"
-	"net"
-	"net/url"
+	gonet "net"
```
Comment (krishnasrinivas) [inline cmd/endpoint.go]: better not change the names of stdlib and instead use minionet:
`minionet "github.com/minio/minio/pkg/net"`

this way we need not worry about which package "net" refers to in our codebase when reading source

Comment (krishnasrinivas) [reply cmd/endpoint.go]: better, we can rename github.com/minio/minio/pkg/net to github.com/minio/minio/pkg/minionet or github.com/minio/minio/pkg/mnet

File: pkg/net/host.go [hunk 5]
```diff
@@ -0,0 +1,137 @@
+package net
```
Comment (poornas) [inline pkg/net/host.go:17]: missing license header

File: pkg/net/port.go [hunk 6]
```diff
@@ -0,0 +1,59 @@
+package net
```
Comment (poornas) [inline pkg/net/port.go:17]: missing license header

File: pkg/net/url.go [hunk 7]
```diff
@@ -0,0 +1,73 @@
+package net
```
Comment (poornas) [inline pkg/net/url.go:17]: license header

**Review by poornas — CHANGES_REQUESTED** [review]

File: cmd/gateway-gcs.go [hunk 8]
```diff
@@ -156,6 +137,63 @@ func (g *GCSGateway) NewGatewayLayer() (GatewayLayer, error) {
...
+	}
+
+	// Handle common command args.
+	handleCommonCmdArgs(ctx)
+
+	var projectID string
+	if len(ctx.Args()) == 1 {
+		projectID = strings.TrimSpace(ctx.Args().First())
+		if !isValidGCSProjectIDFormat(projectID) {
+			fatalIf(errors.New("project ID format error"), "Invalid command line arguments %v", ctx.Args())
```
Comment (poornas) [inline cmd/gateway-gcs.go:164]: "project ID format error" could be a more generic and actionable message

File: cmd/gateway-gcs.go [hunk 9]
```diff
@@ -156,6 +137,63 @@ func (g *GCSGateway) NewGatewayLayer() (GatewayLayer, error) {
...
+		}
+		googleCreds := make(map[string]string)
+		if err = json.NewDecoder(file).Decode(&googleCreds); err != nil {
+			fatalIf(err, "unable to parse JSON in open GOOGLE_APPLICATION_CREDENTIALS file %v", credFile)
+		}
+		file.Close()
+
+		projectID = googleCreds[gcsProjectIDKey]
+		if !isValidGCSProjectIDFormat(projectID) {
+			fatalIf(errors.New("project ID format error"), "invalid value of JSON field %v in GOOGLE_APPLICATION_CREDENTIALS file %v", gcsProjectIDKey, credFile)
```
Comment (poornas) [inline cmd/gateway-gcs.go:190]: "project ID format error" - same feedback as above.

File: cmd/gateway-main.go [hunk 10]
```diff
@@ -223,15 +217,12 @@ func startGateway(ctx *cli.Context, gw Gateway) {
...
-		mode := globalMinioModeGatewayPrefix + gatewayName
-		// Check update mode.
-		checkUpdate(mode)
+	printGatewayStartupMessage(getAPIEndpoints(globalServerHost.String()), gw.Name())
 
-		// Print gateway startup message.
-		printGatewayStartupMessage(getAPIEndpoints(gatewayAddr), gatewayName)
-	}
+	// Set uptime time after object layer has initialized.
+	globalBootTime = UTCNow()
```
Comment (poornas) [inline cmd/gateway-main.go:226]: globalBootTime is not referenced anywhere else. Is this intentional?

**Review by poornas — CHANGES_REQUESTED** [review]

**Comment by harshavardhana** [comment]
Can you rebase this @balamurugana ?

**Comment by kannappanr** [comment]
ping @balamurugana

**Comment by deekoder** [comment]
@balamurugana  please do rebase again because reviewers do look at the code visually as well as test it. We will discuss this rebase policy in our meeting. Alternatively we could wait for the discussion on thu and then take action on this PR. Until then its not actionable to reviewers.

**Pull request closed without merging**